### ISSUE TITLE
max_pool3d does not support float16(half)

### ISSUE THREAD

**Issue by weiliu620** [tensorflow/tensorflow#7632]
I'm using version 0.12.1 on Ubuntu 14.04 LTS 64 bit. Here is the minimal example I run to find the error:

```
    with tf.Graph().as_default():
        a = tf.constant(1, shape = [1, 4, 4, 4, 1], dtype = tf.float16)
        b = tf.nn.max_pool3d(a, ksize = [1, 2, 2, 2, 1], strides = [1, 2, 2, 2, 1], padding = 'SAME')

        with tf.Session() as sess:
            print(sess.run([a, b]))
```

The I got error:

```
...
InvalidArgumentError (see above for traceback): No OpKernel was registered to support Op 'MaxPool3D' with these attrs.  Registered devices: [CPU,GPU], Registered kernels:
  device='CPU'; T in [DT_FLOAT]
  device='GPU'; T in [DT_FLOAT]
```

If I use ```dtype = tf.float32```, there is no error. ```max_pool3d``` python docstring says it supports half precision, but why ```dtype = tf.float16``` does not work?

**Comment by aselle** [comment]
@martinwicke , it is confusing the doc generator gives all the possible types of tensors all the time, even when we don't have kernels registered.

@weiliu620 , thanks for reporting this. We should definitely at least update the documentation. If you would like to provide float16 functionality for maxpool3d that would be great. But we are likely not to get to it anytime soon. Thanks!

**Comment by rilut** [comment]
@aselle greetings, which files are relevant to add float16 support for `maxpool3d`?

**Comment by yaroslavvb** [comment]
It is not beginner friendly, the best way is to find a commit that added fp16 support and use that as a template. Here's a mega-commit that added fp16 to a bunch of ops, although there are probably smaller examples -- https://github.com/tensorflow/tensorflow/commit/9bedadce

**Comment by rilut** [comment]
@yaroslavvb thanks, I'm planning to working on it. Here's my progress so far

**Comment by rilut** [comment]
@yaroslavvb so I compiled my fork of tensorflow (on a Ubuntu VM), and I have compilation errors. Since the error log is long, do you have any tip to debug tf compilation errors?

**Comment by tfboyd** [comment]
We are actively working on FP16.  Assigning to myself to monitor and contributions are still very welcome.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
@MarkDaoust see https://github.com/tensorflow/tensorflow/issues/7632#issuecomment-280898121. 

Could we fix the C++ doc generator to only produce docs for kernels that actually exist? That may not be straightforward, the information exists in different places. @josh11b FYI.

**Comment by MarkDaoust** [comment]
The docs part of this problem is fixed, at some point between r1.1 and r1.2. Compare:

https://www.tensorflow.org/versions/r1.1/api_docs/python/tf/nn/max_pool3d
https://www.tensorflow.org/versions/r1.2/api_docs/python/tf/nn/max_pool3d

I don't know the details of how that list is generated, but given that this is a generated python wrapper that the docs are generated from, I doubt it is something trivial enough that we would be able to cherry-pick back onto the old releases. 

It's not a total solution to your problem, but this op did gain `bfloat16` suport in `r1.5`:

https://www.tensorflow.org/versions/r1.5/api_docs/python/tf/nn/max_pool3d

**Comment by tatianashp** [comment]
The docs issue has been resolved and r1.5 has bfloat16 support.
@rilut Do you still need float16 support and interested in implementing it?

**Comment by rilut** [comment]
@tatianashp Hi, I don't think I really understand the underlying code, so I pass 🙏. Thanks for asking

**Comment by tatianashp** [comment]
I am closing the issue then. If somebody needs fp16 for maxpool3d  and wants to implement please feel free to submit PR.
